syntax = "proto3";

package python;
option go_package = "./gen";

service Python {
  rpc Transcribe(TranscribeRequest) returns (TranscribeResponse);
  rpc Summarize(SummarizeRequest) returns (SummarizeResponse);
  rpc YoutubeTranscript(Video) returns (Transcript);
  rpc Normalize(Text) returns (Text);
  rpc Categorize(CategorizeRequest) returns (Categories);
  rpc IndexDirectory(IndexDirectoryRequest) returns (Index);
  rpc QueryIndex(IndexQuery) returns (QueryResult);
}

enum IndexType {
  LLAMA = 0;
  FAISS = 1;
  BM25 = 2;
}

message IndexQuery {
  string index = 1;
  string query = 2;
}

message QueryResult {
  repeated string results = 1;
}

message IndexDirectoryRequest {
  string path = 1;
  IndexType type = 2;
}

message Index {
  string id = 1;
  IndexType type = 2;
}

message Text {
  string text = 1;
}

message Categories {
  repeated string categories = 1;
}

enum Categorizer {
  T5_TAG = 0;
  KEYBERT = 1;
}

message CategorizeRequest {
  string text = 1;
  Categorizer categorizer = 2;
}

enum Summarizer {
  LANGCHAIN = 0;
  BERT = 1;
}

message SummarizeRequest {
  string content = 1;
  Summarizer summarizer = 2;
}

message SummarizeResponse {
  string summary = 1;
}

message TranscribeRequest {
  string file = 1;
}

message TranscribeResponse {
  string transcription = 1;
}

message Video {
  string id = 1;
}

message TranscriptSection
{
  string text = 1;
  float start = 2;
  float duration = 3;
}

message Transcript {
  repeated TranscriptSection transcript = 1;
}
